pkg/network/tracer/gateway_lookup.go

```diff
@@ -0,0 +1,132 @@
+// +build linux_bpf
+
+package tracer
+
+import (
+	"net"
+	"time"
+	"unsafe"
+
+	ddconfig "github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/network"
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	"github.com/DataDog/datadog-agent/pkg/network/ebpf/probes"
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/ec2"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/ebpf"
+	"github.com/DataDog/ebpf/manager"
+)
+
+const maxRouteCacheSize = int(^uint(0) >> 1) // max int
+
+type gatewayLookup struct {
+	routeCache          network.RouteCache
+	subnetCache         map[int]network.Subnet // interface index to subnet map
+	subnetForHwAddrFunc func(net.HardwareAddr) (network.Subnet, error)
+
+	logLimiter *util.LogLimit
+}
+
+func gwLookupEnabled(config *config.Config) bool {
+	// only enabled on AWS currently
+	return config.EnableGatewayLookup && ddconfig.IsCloudProviderEnabled(ec2.CloudProviderName)
+}
+
+func newGatewayLookup(config *config.Config, runtimeCompilerEnabled bool, m *manager.Manager) *gatewayLookup {
+	if !gwLookupEnabled(config) {
+		return nil
+	}
+
+	var router network.Router
+	if runtimeCompilerEnabled {
+		router = newEbpfRouter(m)
+	} else {
+		router = network.NewNetlinkRouter(config.ProcRoot)
+	}
+
+	routeCacheSize := maxRouteCacheSize
+	if config.MaxTrackedConnections <= uint(maxRouteCacheSize) {
+		routeCacheSize = int(config.MaxTrackedConnections)
+	} else {
+		log.Warnf("using truncated route cache size of %d instead of %d", routeCacheSize, config.MaxTrackedConnections)
+	}
+
+	return &gatewayLookup{
+		subnetCache:         make(map[int]network.Subnet),
+		routeCache:          network.NewRouteCache(routeCacheSize, router),
+		subnetForHwAddrFunc: ec2SubnetForHardwareAddr,
+		logLimiter:          util.NewLogLimit(10, 10*time.Minute),
+	}
+}
+
+func (g *gatewayLookup) Lookup(cs *network.ConnectionStats) *network.Via {
+	r, ok := g.routeCache.Get(cs.Source, cs.Dest, cs.NetNS)
```

Comment:
possible `nil` dereference if `router` is nil in the constructor, and thus `routeCache` is nil.